src/ops/logical.rs

```diff
@@ -0,0 +1,554 @@
+    /// ```
+    ///
+    /// # Params
+    ///
+    /// - other: array to compare
+    /// - stream: stream or device to evaluate on
+    #[default_device]
+    pub fn ne_device(&self, other: &Array, stream: StreamOrDevice) -> Array {
+        unsafe {
+            Array::from_ptr(mlx_sys::mlx_not_equal(
+                self.c_array,
+                other.c_array,
+                stream.as_ptr(),
+            ))
+        }
+    }
+
+    /// Element-wise less than.
+    ///
+    /// Less than on two arrays with <doc:broadcasting>.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use mlx::Array;
+    /// let a = Array::from_slice(&[1, 2, 3], &[3]);
+    /// let b = Array::from_slice(&[1, 2, 3], &[3]);
+    /// let mut c = a.lt_device(&b, Default::default());
+    ///
+    /// c.eval();
+    /// let c_data: &[bool] = c.as_slice();
+    /// // c_data == [false, false, false]
+    /// ```
+    ///
+    /// # Params
+    ///
+    /// - other: array to compare
+    /// - stream: stream or device to evaluate on
+    #[default_device]
+    pub fn lt_device(&self, other: &Array, stream: StreamOrDevice) -> Array {
+        unsafe {
+            Array::from_ptr(mlx_sys::mlx_less(
+                self.c_array,
+                other.c_array,
+                stream.as_ptr(),
+            ))
+        }
+    }
+
+    /// Element-wise greater than.
+    ///
+    /// Greater than on two arrays with <doc:broadcasting>.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use mlx::Array;
+    /// let a = Array::from_slice(&[1, 2, 3], &[3]);
+    /// let b = Array::from_slice(&[1, 2, 3], &[3]);
+    /// let mut c = a.gt_device(&b, Default::default());
+    ///
+    /// c.eval();
+    /// let c_data: &[bool] = c.as_slice();
+    /// // c_data == [false, false, false]
+    /// ```
+    ///
+    /// # Params
+    ///
+    /// - other: array to compare
+    /// - stream: stream or device to evaluate on
+    #[default_device]
+    pub fn gt_device(&self, other: &Array, stream: StreamOrDevice) -> Array {
+        unsafe {
+            Array::from_ptr(mlx_sys::mlx_greater(
+                self.c_array,
+                other.c_array,
+                stream.as_ptr(),
+            ))
+        }
+    }
+
+    /// Element-wise logical and.
+    ///
+    /// Logical and on two arrays with <doc:broadcasting>.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use mlx::Array;
+    /// let a = Array::from_slice(&[true, false, true], &[3]);
+    /// let b = Array::from_slice(&[true, true, false], &[3]);
+    /// let mut c = a.logical_and_device(&b, Default::default());
+    ///
+    /// c.eval();
+    /// let c_data: &[bool] = c.as_slice();
+    /// // c_data == [true, false, false]
+    /// ```
+    ///
+    /// # Params
+    ///
+    /// - other: array to compare
+    /// - stream: stream or device to evaluate on
+    #[default_device]
+    pub fn logical_and_device(&self, other: &Array, stream: StreamOrDevice) -> Array {
```

Comment:
We probably want to check not only the shapes and also the dtype as well

Reply:
I've added a check for broadcasting but dtypes don't need to be compared -- mlx works on any two types, it always converts them to bool first.